atc/db/container.go

```diff
@@ -385,26 +385,44 @@ func (container *failedContainer) Handle() string              { return containe
 func (container *failedContainer) WorkerName() string          { return container.workerName }
 func (container *failedContainer) Metadata() ContainerMetadata { return container.metadata }
 
-func (container *failedContainer) Destroy() (bool, error) {
-	rows, err := psql.Delete("containers").
-		Where(sq.Eq{
-			"id":    container.id,
-			"state": atc.ContainerStateFailed,
+func (container *failedContainer) Destroy() (DestroyingContainer, error) {
+	result, err := psql.Update("containers").
+		Set("state", atc.ContainerStateDestroying).
+		Where(sq.And{
+			sq.Eq{"id": container.id},
+			sq.Or{
+				sq.Eq{"state": string(atc.ContainerStateDestroying)},
+				sq.Eq{"state": string(atc.ContainerStateFailed)},
+				sq.Eq{"state": string(atc.ContainerStateCreated)},
+			},
 		}).
+		PlaceholderFormat(sq.Dollar).
 		RunWith(container.conn).
 		Exec()
 	if err != nil {
-		return false, err
+		if err == sql.ErrNoRows {
```

Comment:
I'm not sure if `Exec()` ever returns `ErrNoRows`? Might want to double check though but I'm _pretty_ sure it doesn't?? But please correct me if I'm wrong.